libkbfs/unflushed_path_cache.go

```diff
@@ -0,0 +1,321 @@
+			len(newUnflushedPaths))
+	}
+
+	return newUnflushedPaths[irmd.Revision()], nil
+}
+
+// appendToCache returns true when successful, and false if it needs
+// to be retried after the per-revision map is recomputed.
+func (upc *unflushedPathCache) appendToCache(irmd ImmutableRootMetadata,
+	perRevMap unflushedPathsPerRevMap) bool {
+	upc.lock.Lock()
+	defer upc.lock.Unlock()
+	switch upc.state {
+	case upcUninitialized:
+		// Nothing to do.
+	case upcInitializing:
+		// Append to queue for processing at the end of initialization.
+		upc.appendQueue = append(upc.appendQueue, irmd)
+	case upcInitialized:
+		if perRevMap == nil {
+			// This was prepared before `upc.chainsPopulator` was set,
+			// and needs to be done again.
+			return false
+		}
+		// Update the cache with the prepared paths.
+		upc.unflushedPaths[irmd.Revision()] = perRevMap
+	default:
+		panic(fmt.Sprintf("Unknown unflushedPathsCache state: %v", upc.state))
+	}
+	return true
+}
+
+func (upc *unflushedPathCache) removeFromCache(rev MetadataRevision) {
+	upc.lock.Lock()
+	defer upc.lock.Unlock()
+	switch upc.state {
+	case upcUninitialized:
+		// Nothing to do.
+	case upcInitializing:
+		// Append to queue for processing at the end of initialization.
+		upc.removeQueue = append(upc.removeQueue, rev)
+	case upcInitialized:
+		delete(upc.unflushedPaths, rev)
+	default:
+		panic(fmt.Sprintf("Unknown unflushedPathsCache state: %v", upc.state))
+	}
+}
+
+func (upc *unflushedPathCache) setCacheIfPossible(cache unflushedPathsMap,
+	cpp chainsPathPopulator) []ImmutableRootMetadata {
+	upc.lock.Lock()
+	defer upc.lock.Unlock()
+	if len(upc.appendQueue) > 0 {
+		// We need to process more appends!
+		queue := upc.appendQueue
+		upc.appendQueue = nil
+		return queue
+	}
+
+	for _, rev := range upc.removeQueue {
+		delete(cache, rev)
+	}
+	upc.removeQueue = nil
+	upc.unflushedPaths = cache
+	upc.chainsPopulator = cpp
+	close(upc.ready)
+	upc.ready = nil
+	upc.state = upcInitialized
+	return nil
+}
+
+func (upc *unflushedPathCache) initialize(ctx context.Context,
+	uid keybase1.UID, kid keybase1.KID, codec kbfscodec.Codec,
+	log logger.Logger, cpp chainsPathPopulator,
+	irmds []ImmutableRootMetadata) (unflushedPathsMap, error) {
+	// First get all the paths for the given range.  On the first try
+	unflushedPaths := make(unflushedPathsMap)
+	log.CDebugf(ctx, "Initializing unflushed path cache with %d revisions",
+		len(irmds))
+	err := addUnflushedPaths(ctx, uid, kid, codec, log, irmds, cpp,
+		unflushedPaths)
+	if err != nil {
+		return nil, err
+	}
+
+	initialUnflushedPaths := make(unflushedPathsMap)
+	// Only need to deep-copy the outer level of the map; the inner
+	// level (per-revision) is immutable.
+	for k, v := range unflushedPaths {
+		initialUnflushedPaths[k] = v
+	}
+
+	// This loop is limited only by the lifetime of `ctx`.  TODO: we
+	// might want to add a way to block appendToCache calls if we loop
+	// too many times here waiting for an empty queue.
+	for {
```

Comment:
All the unflushed path computation is purely informational, right? It seems to me that journal ops should take precedence over the unflushed path computation. What about returning partial results, since we already do that anyway? (And reverting back to uninitialized state.)